selfdrive/car/ford/radar_interface.py

```diff
@@ -1,72 +1,99 @@
-                     RADAR_MSGS * 3))
-  checks = list(zip(RADAR_MSGS, [20] * msg_n))
+def _create_radar_can_parser(CP):
+  if DBC[CP.carFingerprint]['radar'] is None:
+    return None
 
-  return CANParser(DBC[car_fingerprint]['radar'], signals, checks, CANBUS.radar)
+  signals = []
+  checks = []
+
+  for i in range(1, NUM_MSGS + 1):
+    msg = f"MRR_Detection_{i:03d}"
+    signals += [
+      (f"CAN_DET_VALID_LEVEL_{i:02d}", msg),
+      (f"CAN_DET_RANGE_{i:02d}", msg),
+      (f"CAN_DET_AZIMUTH_{i:02d}", msg),
+      (f"CAN_DET_RANGE_RATE_{i:02d}", msg),
+      (f"CAN_DET_AMPLITUDE_{i:02d}", msg),
+      (f"CAN_SCAN_INDEX_2LSB_{i:02d}", msg),
+    ]
+    checks += [(msg, 20)]
+
+  return CANParser(DBC[CP.carFingerprint]['radar'], signals, checks, CANBUS.radar)
 
 class RadarInterface(RadarInterfaceBase):
   def __init__(self, CP):
     super().__init__(CP)
-    self.validCnt = {key: 0 for key in RADAR_MSGS}
+    self.updated_messages = set()
+    self.trigger_msg = LAST_MSG
     self.track_id = 0
 
-    self.rcp = _create_radar_can_parser(CP.carFingerprint)
-    self.trigger_msg = 0x53f
-    self.updated_messages = set()
+    self.rcp = _create_radar_can_parser(CP)
 
   def update(self, can_strings):
+    if self.rcp is None:
+      return super().update(None)
+
     vls = self.rcp.update_strings(can_strings)
     self.updated_messages.update(vls)
 
     if self.trigger_msg not in self.updated_messages:
       return None
 
+    rr = self._update(self.updated_messages)
+    self.updated_messages.clear()
+
+    return rr
+
+  def _update(self, updated_messages):
     ret = car.RadarData.new_message()
+    if self.rcp is None:
+      return ret
+
     errors = []
+
     if not self.rcp.can_valid:
       errors.append("canError")
     ret.errors = errors
 
-    for ii in sorted(self.updated_messages):
-      cpt = self.rcp.vl[ii]
+    for i in range(1, NUM_MSGS + 1):
+      msg = self.rcp.vl[f"MRR_Detection_{i:03d}"]
+      index = msg[f"CAN_SCAN_INDEX_2LSB_{i:02d}"]
+      ii = (i - 1) * 4 + index
 
-      if cpt['X_Rel'] > 0.00001:
-        self.validCnt[ii] = 0    # reset counter
+      if ii not in self.pts:
+        self.pts[ii] = car.RadarData.RadarPoint.new_message()
+        self.pts[ii].trackId = self.track_id
+        self.track_id += 1
+
+      # radar point only valid if valid signal asserted
+      valid = msg[f"CAN_DET_VALID_LEVEL_{i:02d}"] > 0
+      if valid:
+        rel_distance = msg[f"CAN_DET_RANGE_{i:02d}"]  # m
+        azimuth = msg[f"CAN_DET_AZIMUTH_{i:02d}"]  # rad
+
+        self.pts[ii].dRel = rel_distance  # m from front of car
+        self.pts[ii].yRel = sin(-azimuth) * rel_distance  # in car frame's y axis, left is positive
+        self.pts[ii].vRel = msg[f"CAN_DET_RANGE_RATE_{i:02d}"]  # m/s relative velocity
+
+        # use aRel for debugging AMPLITUDE (reflection size)
+        self.pts[ii].aRel = msg[f"CAN_DET_AMPLITUDE_{i:02d}"]  # dBsm
```

Comment:
I'm not sure where it's gone wrong but `aRel` always seems to be zero when it gets to `tools/replay/ui.py`...

Reply:
There is a limit to its range in cereal